kratos/solving_strategies/builder_and_solvers/residualbased_elimination_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1352 @@
+
+                // Gets list of Dof involved on every element
+                it_const->GetDofList(dof_list, auxiliar_dof_list, r_current_process_info);
+                dofs_aux_list_all[this_thread_id].insert(dof_list.begin(), dof_list.end());
+                dofs_aux_list_all[this_thread_id].insert(auxiliar_dof_list.begin(), auxiliar_dof_list.end());
+            }
+        }
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 2)) << "Initializing tree reduction\n" << std::endl;
+
+        // Here we do a reduction in a tree so to have everything on thread 0
+        // TODO: move to a function
+        /* For all DoF */
+        IndexType old_max = nthreads;
+        IndexType new_max = ceil(0.5*static_cast<double>(old_max));
+        while (new_max>=1 && new_max != old_max) {
+            // Just for debugging
+            KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", this->GetEchoLevel() > 2) << "old_max" << old_max << " new_max:" << new_max << std::endl;
+            for (int i = 0; i < static_cast<int>(new_max); ++i) {
+                KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", this->GetEchoLevel() > 2 && (i + new_max < old_max)) << i << " - " << i+new_max << std::endl;
+            }
+
+            #pragma omp parallel for
+            for (int i = 0; i < static_cast<int>(new_max); ++i) {
+                if (i + new_max < old_max) {
+                    dofs_aux_list_all[i].insert(dofs_aux_list_all[i+new_max].begin(), dofs_aux_list_all[i+new_max].end());
+                    dofs_aux_list_all[i+new_max].clear();
+                }
+            }
+
+            old_max = new_max;
+            new_max = ceil(0.5*static_cast<double>(old_max));
+        }
+
+        KRATOS_INFO_IF("ResidualBasedEliminationBuilderAndSolverWithConstraints", ( this->GetEchoLevel() > 2)) << "Initializing ordered array filling\n" << std::endl;
+
+        dof_temp_all.reserve(dofs_aux_list_all[0].size());
+        for (auto& dof : dofs_aux_list_all[0]) {
+            dof_temp_all.push_back( dof.get() );
+        }
+        dof_temp_all.Sort();
+        BaseType::mDofSet = dof_temp_all;
+
+        // Repeating the loop for solvable DoF (in serial)
+        set_type dofs_aux_list_solvable;
+        // Gets the array of elements from the modeler
+        ElementsArrayType& r_elements_array = rModelPart.Elements();
+        const int number_of_elements = static_cast<int>(r_elements_array.size());
+        #pragma omp for schedule(guided, 512) nowait
+        for (int i = 0; i < number_of_elements; ++i) {
+            auto it_elem = r_elements_array.begin() + i;
+
+            // Gets list of Dof involved on every element
+            pScheme->GetElementalDofList(*(it_elem.base()), dof_list, r_current_process_info);
+            dofs_aux_list_solvable.insert(dof_list.begin(), dof_list.end());
```

Comment:
this is not threadsafe

Reply:
It is done in the others builder and solvers